gbn/queue.go

```diff
@@ -197,17 +417,40 @@ func (q *queue) processNACK(seq uint8) (bool, bool) {
 
 	q.cfg.log.Tracef("Received NACK %d", seq)
 
-	// If the NACK is the same as sequenceTop, it probably means that queue
+	if q.awaitingCatchUp && seq == q.awaitedNACK {
+		q.cfg.log.Tracef("Sending awaitedNACKSignal")
+		q.awaitedNACKSignal <- struct{}{}
+
+		q.awaitingCatchUp = false
+
+		// In case the awaitedNACK is the same as sequenceTop, we can
+		// bump the base to be equal to sequenceTop, without triggering
+		// a new resend.
+		if seq == q.sequenceTop {
+			q.sequenceBase = q.sequenceTop
+		}
+
+		// If we receive the awaited NACK, we shouldn't trigger a new
+		// resend, as we can now proceed to send new packets.
+		return false, false
+	}
+
+	// If the NACK is the same as sequenceTop, and we weren't awaiting this
+	// NACK as part of the resend catch up, it probably means that queue
 	// was sent successfully, but we just missed the necessary ACKs. So we
-	// can empty the queue here by bumping the base and we dont need to
+	// can empty the queue here by bumping the base and we don't need to
 	// trigger a resend.
 	if seq == q.sequenceTop {
 		q.sequenceBase = q.sequenceTop
-		return true, false
+
+		return false, false
 	}
 
 	// Is the NACKed sequence even in our queue?
 	if !containsSequence(q.sequenceBase, q.sequenceTop, seq) {
+		q.cfg.log.Tracef("NACK seq %d is not in the queue. Ignoring.",
```

Comment:
it seems the "bumped" logic is changed here only to be changed again in a later commit...

If there is a bug in the original logic - perhaps better to fix that in it's own commit _before_ any additions. Cause these 2 diffs are a bit confusing to follow

Reply:
Yeah good idea, cleaned this up with the latest push :)!